File: client/src/models/BookInRankingViewModel.ts
import { z } from 'zod';
import PublisherViewModel from './PublisherViewModel';
import AuthorViewModel from './author/AuthorViewModel';
import BookCategoryViewModel from './BookCategoryViewModel';

const BookInRankingViewModel = z.object({
  id: z.string().uuid(),
  isbn: z.string().max(17),
  title: z.string(),
  coverPictureUrl: z.string().url().max(256).nullable(),
  averageRating: z.nullable(z.number()),
  averageCriticRating: z.nullable(z.number()),
  publisher: PublisherViewModel.nullable(),
  authors: z.array(AuthorViewModel),
  bookCategories: z.array(BookCategoryViewModel),
  reviewsCount: z.number().int(),
});

export default BookInRankingViewModel;
export type BookInRankingViewModelType = z.infer<typeof BookInRankingViewModel>;

File: Server/src/Server.Application/ViewModels/BookInRankingViewModel.cs
// ReSharper disable CollectionNeverUpdated.Global
#pragma warning disable CS8618
namespace Server.Application.ViewModels;

public class BookInRankingViewModel
{
    public Guid Id { get; set; }
    public string ISBN { get; set; }
    public string Title { get; set; }
    public double? AverageRating { get; set; }
    public double? AverageCriticRating { get; set; }
    public string? CoverPictureUrl { get; set; }
    public ICollection<AuthorViewModel> Authors { get; set; }
    public PublisherViewModel Publisher { get; set; }
    public ICollection<BookCategoryViewModel> BookCategories { get; set; }
    public int ReviewsCount { get; set; }
    public int CriticReviewsCount { get; set; }
}